filament/backend/src/vulkan/VulkanAsyncHandles.h

```diff
@@ -70,28 +74,43 @@ struct VulkanFence : public HwFence, fvkmemory::ThreadSafeResource {
     VulkanFence() {}
 
     void setFence(std::shared_ptr<VulkanCmdFence> fence) {
-        std::lock_guard lock(mState->lock);
+        std::lock_guard const lock(mState->lock);
         mState->sharedFence = std::move(fence);
         mState->cond.notify_all();
     }
 
     std::shared_ptr<VulkanCmdFence>& getSharedFence() {
-        std::lock_guard lock(mState->lock);
+        std::lock_guard const lock(mState->lock);
         return mState->sharedFence;
     }
 
-    std::shared_ptr<VulkanCmdFence> wait(std::chrono::steady_clock::time_point const until) {
+    std::pair<std::shared_ptr<VulkanCmdFence>, FenceStatus>
+            wait(std::chrono::steady_clock::time_point const until) {
         // hold a reference so that our state doesn't disappear while we wait
         std::shared_ptr state{ mState };
         std::unique_lock lock(state->lock);
-        state->cond.wait_until(lock, until, [&state] { return bool(state->sharedFence); });
+        state->cond.wait_until(lock, until, [&state] {
+            return bool(state->sharedFence) || state->status != FenceStatus::TIMEOUT_EXPIRED;
+        });
         // here mSharedFence will be null if we timed out
-        return state->sharedFence;
+        return { state->sharedFence, state->status };
     }
+
+    void cancel() const {
+        std::shared_ptr const state{ mState };
+        std::unique_lock const lock(state->lock);
+        if (state->sharedFence) {
+            state->sharedFence->cancel();
+        }
+        state->status = FenceStatus::ERROR;
+        state->cond.notify_all();
+    }
+
 private:
     struct State {
         std::mutex lock;
         std::condition_variable cond;
+        FenceStatus status = FenceStatus::TIMEOUT_EXPIRED;
```

Comment:
We don't really need the additional semantics of FenceStatus. It'll be easier to parse with a well-named boolean like `hasCancelled` IMO.